haystack/components/joiners/document_joiner.py

```diff
@@ -171,3 +174,28 @@ def _reciprocal_rank_fusion(self, document_lists):
             doc.score = scores_map[doc.id]
 
         return documents_map.values()
+
+    def _distribution_based_rank_fusion(self, document_lists):
```

Comment:
Missing type-hint and return type.

Reply:
Excuse me @shadeMe but I don't understand. I followed the structure of the other methods in this class, none have type-hint and return type.

```
def _merge(self, document_lists):
        """
        Merge multiple lists of Documents and calculate a weighted sum of the scores of duplicate Documents.
        """
        scores_map = defaultdict(int)
        documents_map = {}
        weights = self.weights if self.weights else [1 / len(document_lists)] * len(document_lists)

        for documents, weight in zip(document_lists, weights):
            for doc in documents:
                scores_map[doc.id] += (doc.score if doc.score else 0) * weight
                documents_map[doc.id] = doc

        for doc in documents_map.values():
            doc.score = scores_map[doc.id]

        return documents_map.values()
```

Reply:
Ah, that's clearly an oversight w.r.t the other methods. Feel free to also fill in the type hints in the other methods.

Reply:
Done 👍

Reply:
I'm afraid we need to fix both the missing type hints and the linting errors before merging.

Could you please try to reproduce the lint errors locally (you can look up the command on the CI run and ensure that you have the same version of `pylint`/`mypy` installed locally) and fix them?

Reply:
Sorry, at the moment I can't. You could reject the PR of it's necessary.

Reply:
@shadeMe if I try to apply naive typing (`def join_method(self , document_lists: Variadic[List[Document]]) -> List[Document]:`),
I get the following errors:

> haystack/components/joiners/document_joiner.py:132: error: Need type annotation for "scores_map"  [var-annotated]
haystack/components/joiners/document_joiner.py:134: error: Argument 1 to "len" has incompatible type "Iterable[list[Document]]"; expected "Sized"  [arg-type]
haystack/components/joiners/document_joiner.py:144: error: Incompatible return value type (got "dict_values[str, Document]", expected "list[Document]")  [return-value]
haystack/components/joiners/document_joiner.py:155: error: Need type annotation for "scores_map"  [var-annotated]
haystack/components/joiners/document_joiner.py:157: error: Argument 1 to "len" has incompatible type "Iterable[list[Document]]"; expected "Sized"  [arg-type]
haystack/components/joiners/document_joiner.py:162: error: Unsupported operand types for * ("float" and "int")  [operator]
haystack/components/joiners/document_joiner.py:162: error: Argument 1 to "len" has incompatible type "Iterable[list[Document]]"; expected "Sized"  [arg-type]
haystack/components/joiners/document_joiner.py:168: error: Argument 1 to "len" has incompatible type "Iterable[list[Document]]"; expected "Sized"  [arg-type]
haystack/components/joiners/document_joiner.py:173: error: Incompatible return value type (got "dict_values[str, Document]", expected "list[Document]")  [return-value]

so I think that, intentionally or unintentionally, we have always neglected the typing of this component and probably should fix it in another PR.

(If I have time, I might take a look at it).

Reply:
In the end, fixing the typing took less work than expected and I did it in this PR.